flowbench/BenchVersionVectorSerialization.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * BenchVersionVector.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "flow/Arena.h"
+#include "benchmark/benchmark.h"
+#include "fdbclient/VersionVector.h"
+#include <cstdint>
+
+struct TestContextArena {
+	Arena& _arena;
+	Arena& arena() { return _arena; }
+	ProtocolVersion protocolVersion() const { return g_network->protocolVersion(); }
+	uint8_t* allocate(size_t size) { return new (_arena) uint8_t[size]; }
+};
+
+static void bench_serializable_traits_version(benchmark::State& state) {
+	int tagCount = state.range(0);
+
+	Version version = 100000;
+	VersionVector serializedVV(version);
+	for (int i = 0; i < tagCount; i++) {
+		serializedVV.setVersion(Tag(0, i), ++version);
+	}
+
+	size_t size = 0;
+	VersionVector deserializedVV;
+	while (state.KeepRunning()) {
+		Standalone<StringRef> msg = ObjectWriter::toValue(serializedVV, Unversioned());
```

Comment:
It turns out we need to derive "serializable_traits<VersionVector>" from "std::true_type"(in "VersionVector.h") in order to get "ObjectWriter::toValue()" to take the code path that "serializable_traits<VersionVector>::serialize()" takes.

Here is the benchmark output when "serializable_traits<VersionVector>" is derived from "std::false_type" (which is very similar to the output of "bench_dynamic_size_traits_version" version):

> -------------------------------------------------------------------------------------------------
> Benchmark                                       Time             CPU   Iterations UserCounters...
> -------------------------------------------------------------------------------------------------
> bench_serializable_traits_version/16         1343 ns         1348 ns       518613 Size=104 Tags=16 items_per_second=742.101k/s
> bench_serializable_traits_version/64         2353 ns         2358 ns       298759 Size=200 Tags=64 items_per_second=424.073k/s
> bench_serializable_traits_version/512       29690 ns        29703 ns        23644 Size=2.12k Tags=512 items_per_second=33.6662k/s
> bench_serializable_traits_version/1024      64459 ns        64498 ns        10813 Size=4.168k Tags=1024 items_per_second=15.5043k/s

Here is the benchmark output when "serializable_traits<VersionVector>" is derived from "std::true_type" (note the difference in serialized buffer size output between this output and the previous output):

> -------------------------------------------------------------------------------------------------
> Benchmark                                       Time             CPU   Iterations UserCounters...
> -------------------------------------------------------------------------------------------------
> bench_serializable_traits_version/16          772 ns          772 ns       935419 Size=392 Tags=16 items_per_second=1.29541M/s
> bench_serializable_traits_version/64         2189 ns         2189 ns       319626 Size=1.352k Tags=64 items_per_second=456.799k/s
> bench_serializable_traits_version/512       15831 ns        15830 ns        44204 Size=10.312k Tags=512 items_per_second=63.1695k/s
> bench_serializable_traits_version/1024      31625 ns        31625 ns        22147 Size=20.552k Tags=1024 items_per_second=31.6204k/s
> 

Note that "serializable_traits<VersionVector>" is currently derived from "std::false_type" in "main". Shall we modify it to be derived from "std::true_type"?

Reply:
Yes, since the change will make serialization faster. We may revisit this decision when the serialized size becomes a concern.